### ISSUE TITLE
[Feature Request]: Full TLB Support

### ISSUE THREAD

**Issue by TheLastRar** [PCSX2/pcsx2#11910]
### Description

Currently, we use a fixed mapping based on the pre initialised values set on boot.
These mappings can later be changed, however, almost no software made use of this

Only 1 game, and Linux (derived) made use of this feature.

Due to the complexity and minimal usage, there is currently no plan for tacking this feature ourselves.

### Reason

PS2 Linux and PSBBN (also based on Linux) require full TLB emulation.

Goemon also uses a non default TLB mapping (and I think is the only game to do so), but we have a hack for that so isn't a stopper for that.

### Examples

https://github.com/PCSX2/pcsx2/issues/11746
https://github.com/PCSX2/pcsx2/issues/5403
https://github.com/PCSX2/pcsx2/issues/8656

**Comment by Goatman13** [comment]
POPS (official PS1 discless emulator) and PS2PSXe (homebrew PS1 emulator) use custom TLB mappings too.

**Comment by Asd677gg** [comment]
> Sorry for bumping, but ive been searching around countless threads about it on how to successfully boot PSBBN, theres recently exciting updates by people making English patch for it. So making a virtualized PSBBN a la the wonderful Xbox dashboards emulation on XEMU (LLE) sounds exciting.
> 
> Ive been trying my built fork off someones Magicgate fork, but tlb is something im not experienced with to build even if not possible i think so far as only built in tlb to be made available isnt something out of question . So a feature request for it is very valid.

I've also searched a lot of threads on how to successfully run PSBBN. With magicgate fork i only get stuck for a couple of seconds on the playstation 2 logo. After that, it will throw me back to the browser screen.

**Comment by toffee-makes-things** [comment]
> I've also searched a lot of threads on how to successfully run PSBBN. With magicgate fork i only get stuck for a couple of seconds on the playstation 2 logo. After that, it will throw me back to the browser screen.

that's most likely due to Dev9 issues that have already been fixed (see #11746) the main issue now is with TLB @Asd677gg @marxman1412

**Comment by Asd677gg** [comment]
> > I've also searched a lot of threads on how to successfully run PSBBN. With magicgate fork i only get stuck for a couple of seconds on the playstation 2 logo. After that, it will throw me back to the browser screen.
> 
> that's most likely due to Dev9 issues that have already been fixed (see #11746) the main issue now is with TLB @Asd677gg @marxman1412 

You mean that i need to use the latest version of pcsx2 + magicgate and it will work? And also, do i need to decrypt the elf file? If yes, then how?

**Comment by kamfretoz** [comment]
> You mean that i need to use the latest version of pcsx2 + magicgate and it will work? And also, do i need to decrypt the elf file? If yes, then how?

No. the DEV9 related issues (that was fixed a few months back) was one of the two last blockers needed for PSBBN to work. The other main one is being the lack of Full TLB Support that were mentioned.

**Comment by gamerx1-cyber** [comment]
This will vastly improve ps1 compatibility on DECKARD series of consoles (SCPH-7500+) as it uses full TLB to software emulate

**Comment by StillPlay2** [comment]
> This will vastly improve ps1 compatibility on DECKARD series of consoles (SCPH-7500+) as it uses full TLB to software emulate

That's right, on the DECKARDS, they replaced the IOP & used a software approach that utilises full TLB

**Comment by RedPanda4552** [comment]
> This will vastly improve ps1 compatibility on DECKARD series of consoles (SCPH-7500+) as it uses full TLB to software emulate

Is there a citation for this? If the firmware on the Deckard chip emulating the IOP is somehow directly addressing the EE memory space and modifying its TLB layout, that is... Alarming, to put it lightly. I can't help but say this feels like LLM slop being leveraged for nefarious reasons, without a citation to back it up.

**Comment by gamerx1-cyber** [comment]
> > This will vastly improve ps1 compatibility on DECKARD series of consoles (SCPH-7500+) as it uses full TLB to software emulate
> 
> Is there a citation for this? If the firmware on the Deckard chip emulating the IOP is somehow directly addressing the EE memory space and modifying its TLB layout, that is... Alarming, to put it lightly. I can't help but say this feels like LLM slop being leveraged for nefarious reasons, without a citation to back it up.

https://www.copetti.org/writings/consoles/playstation-2/#the-special-cpu

This suggests the IOP & the EE are in direct connection & modifying its mappings when in ps1 mode.

https://www.ps2-home.com/forum/viewtopic.php?t=7968

"It is possible to access PS1 memory in slim PS2 with deckard models. PS1DRV here installs a kernel patch for messing with TLB and edits a existing one to enable EE access to SIF0,1,2 DMA registers and right after it adds a new TLB which maps IOP memory in 0x1C000000. This address can be accessed by EE even in PS1 mode and it serves as address 0 of PS1 memory, the very first byte of the 2MB PS1 memory. Using this information allows alot to be done because the IOP memory can be changed by EE, although there are some limitations for sure. Another way of doing working with IOP is loading code in DECKCARD emulator RAM before switching to PS1 mode. wisi has achieved this amazing thing!!!
Slim PS1DRV use SIF2 for sending and receiving something, however, I don't really know how it works."

**Comment by F0bes** [comment]
> This suggests the IOP & the EE are in direct connection & modifying its mappings when in ps1 mode.

What you sent didn't mention anything about modifying the mappings during PS1 mode.
And yes, the EE has a direct mapping to the IOP, more on that below.

> https://www.ps2-home.com/forum/viewtopic.php?t=7968
> 
> "It is possible to access PS1 memory in slim PS2 with deckard models. PS1DRV here installs a kernel patch for messing with TLB and edits a existing one to enable EE access to SIF0,1,2 DMA registers and right after it adds a new TLB which maps IOP memory in 0x1C000000. This address can be accessed by EE even in PS1 mode and it serves as address 0 of PS1 memory, the very first byte of the 2MB PS1 memory. Using this information allows alot to be done because the IOP memory can be changed by EE, although there are some limitations for sure. Another way of doing working with IOP is loading code in DECKCARD emulator RAM before switching to PS1 mode. wisi has achieved this amazing thing!!!
> Slim PS1DRV use SIF2 for sending and receiving something, however, I don't really know how it works."

SIF 0,1 and 2 FIFOS are available to the EE, not really exclusive to Deckard or PS1 mode.

The IOP memory mapping at 0x1C000000 is already supported by PCSX2.

I fail to see what we are missing w.r.t. EE TLB and PS1 mode.

**Comment by Goatman13** [comment]
> I fail to see what we are missing w.r.t. EE TLB and PS1 mode.

New ps1drv writes to 0x1F8014xx / 0x1F8015xx page on EE side. I guess that's custom mapped.

> This will vastly improve ps1 compatibility on DECKARD series of consoles (SCPH-7500+) as it uses full TLB to software emulate

Ps1 driver from consoles with PPC use different IO ports for emulation. For example, on old driver pgpu stat is at 0x1000F300, on new is at 0x1000F200 (MSCOM port) .  Emulated PS1 GPU E2 register on old version is at 0x1000F310 while on new version it was moved at 0x1F801504. PCSX2 tries to emulate old version and expect few things to be on addresses that are wrong for new driver. There is much more to do than map TLB to make it work.

**Comment by byteshiftlabs** [comment]
Hello,

This would be my first contribution to PCSX2, so before I put time into it: is anyone currently working on full TLB support? Didn't want to duplicate effort if so.

I saw in #8656 that stenzek had a WIP branch with full TLB emulation, interpreter-only, that failed once it reached userspace. That branch/fork doesn't seem to exist anymore, so I can't look at what was there.

I'm planning to work through this using ps2tek's documentation, along with AI assistance for research. If there are any other doc resources worth looking at, please let me know. Any feedback is appreciated 😃

**Comment by maekawa-mugi** [comment]
Regarding the full TLB implementation, I was able to reach user space when it was implemented correctly (with AI assistance, obviously). However, I cannot submit a pull request because I lack the necessary domain knowledge to contribute to PCSX2 development and haven't yet fully grasped the source tree well enough to conduct a review. Furthermore, the recompiler implementation is significantly more complex; I question whether it is worth maintaining such complex code for the sake of just a single piece of software.

**Comment by byteshiftlabs** [comment]
Fair enough! I thought of tackling the interpreter first for simplicity, precisely for that reason you mentioned. As I'm new here, and given the complexity of it and the very few use cases it covers, I'd rather let the community weigh in on this👍🏻

**Comment by Florin9doi** [comment]
#14844

**Comment by Arawn-Davies** [comment]
> [#14844](https://github.com/PCSX2/pcsx2/issues/14844)

yeah I got bored
heard you guys like Linux, so I gave you Linux inside PCSX2 inside Linux inside Windows

**Comment by Arawn-Davies** [comment]
I've drafted a PR, not finished (I mean it's functionally finished but by no means ready to merge. Needs more dissecting and cleaning up comments, other AI hallucinations etc.). 

Basically, in laymans terms, it's not making a full note of which virtual memory addresses match with which physical memory addresses, which are read-only, and some BIOS HLE syscall stuff (zero-based array indexing vs 1-based array indexing). 

I don't wanna leave it hard-to-read, comment-less and overly complex, but if anyone want's to take a look at it and send me a diff or .patch, I think we can get this baby bootin' 😂

https://github.com/PCSX2/pcsx2/pull/14847/changes

Also this only affects the interpreter, recompiler is a whole other tin of worms which terrifies me to bits lol

**Comment by Arawn-Davies** [comment]
@maekawa-mugi may I ask why you down voted the comment I put? I'm trying to be helpful and get it working.

**Comment by Arawn-Davies** [comment]
@GovanifY @maekawa-mugi I wasn't demanding a review, it was a starting point. I'm sorry if I came across rude as it was not my intention.
